functions/Rename-DbaLogin.ps1

```diff
@@ -116,46 +123,51 @@ function Rename-DbaLogin {
 
-            foreach ($db in $dbenums) {
-                $db = $databases[$db.DBName]
-                $user = $db.Users[$Login]
-                Write-Message -Level Verbose -Message "Starting update for $db"
-
-                if ($Pscmdlet.ShouldProcess($SqlInstance, "Changing database $db user $user from [$Login] to [$NewLogin]")) {
-                    try {
-                        $oldname = $user.name
-                        $user.Rename($NewLogin)
-                        [pscustomobject]@{
-                            ComputerName = $server.ComputerName
-                            InstanceName = $server.ServiceName
-                            SqlInstance  = $server.DomainInstanceName
-                            Database     = $db.name
-                            PreviousUser = $oldname
-                            NewUser      = $NewLogin
-                            Status       = "Successful"
+            if ($Force) {
+                foreach ($db in $dbenums) {
+                    $db = $databases[$db.DBName]
+                    $user = $db.Users[$Login]
+                    if ($user) {
+                        Write-Message -Level Verbose -Message "Starting update for $db"
+
+                        if ($Pscmdlet.ShouldProcess($SqlInstance, "Changing database $db user $user from [$Login] to [$NewLogin]")) {
+                            try {
+                                $oldname = $user.name
+                                $null = $user.Rename($NewLogin)
+                                $output += [pscustomobject]@{
+                                    ComputerName = $server.ComputerName
+                                    InstanceName = $server.ServiceName
+                                    SqlInstance  = $server.DomainInstanceName
+                                    Database     = $db.name
+                                    PreviousUser = $oldname
+                                    NewUser      = $NewLogin
+                                    Status       = "Successful"
+                                }
+                            } catch {
+                                Write-Message -Level Warning -Message "Rolling back update to login: $Login"
+                                $currentLogin.rename($Login)
+
+                                [pscustomobject]@{
+                                    ComputerName = $server.ComputerName
+                                    InstanceName = $server.ServiceName
+                                    SqlInstance  = $server.DomainInstanceName
+                                    Database     = $db.name
+                                    PreviousUser = $NewLogin
+                                    NewUser      = $oldname
+                                    Status       = "Failure to rename. Rolled back change."
+                                }
+                                Stop-Function -Message "Failure" -ErrorRecord $_ -Target $NewLogin -Continue
```

Comment:
This `-Continue` will not leave the `foreach ($db in $dbenums) {` but go on with the next database user. You will have to use a `break` to leave the loop and continue with the next $instance.

Reply:
good call. i was told by Fred not to use break anywhere as it breaks someone entire pipeline, so we use return instead.

Reply:
oh yes, return is best here.

Reply:
You don't want to return in a foreach loop, that breaks the whole loop. Continue is correct method as it moves onto the next database in the loop.

Reply:
Processing side if we fail in one database for remapping a user it is a matter if you want it to fail for all of them or keep trying.

As it stands this will tell it to keep trying, but the rest will fail if we are going to roll back the login rename. This may cause a number of random errors. Only reason the catch could get hit is if something changed after 104 when we got the `$dbenums`. The other issue would be permissions, the user had rights to rename the login but not touch any databases (e.g. giving [securityadmin ](https://docs.microsoft.com/en-us/sql/relational-databases/security/authentication-access/server-level-roles?view=sql-server-ver15#fixed-server-level-roles) but denied access to databases)

Reply:
Ok, I was not correct with the "return is best" as it ends the command even if the failure is in the first of more instances to process. But "continue" on the other hand would rename other database users, while the login rename is rolled back.
Suggestion: Forget about rolling back (as you don't roll back the already renamed database users anyway) and just try to rename the user in every database and report the failed ones. So the gereral rule is (I think it's the same in other commands): We do as much as we can, and report if we failed - so the DBA has to do the rest.